### ISSUE TITLE
Unexpected behaviour of `ParseResults.copy()` with nested ParseResults

### ISSUE THREAD

**Issue by brynpickering** [pyparsing/pyparsing#463]
I'm building a parsing grammar similar to that given in [eval_arith.py](https://github.com/pyparsing/pyparsing/blob/master/examples/eval_arith.py), with some extensions. However, ideally, one step in the process involves searching a copy of the parsed object and replacing some elements (i.e., leaving the original parsed object intact).

This doesn't work as copying the `ParseResults` object does not fully decouple it. I've tried multiple ways to get a fully decoupled copy (see below) and none seem to work. There are two issues in the behaviour of `copy()` that are unexpected:

1. Changing a nested element of a copy also changes the original. Maybe there needs to be an option for `copy()` to force this, e.g., `recursive` or `deep`.
2. Changing a copy of a nested element inside a copy of the top-level object changes the original. This seems to be a bug.

The same behaviour can be seen with `comp_expr` in `eval_arith.py`, so I will use that for an example:

```python
parsed_ = comp_expr.parse_string("1 * 2 = 2", parse_all=True)
print(parsed_.dump())
print(parsed_[0].value.dump())
```
Output:
```python
[<__main__.EvalComparisonOp object at 0x165f3ef40>]
[<__main__.EvalMultOp object at 0x165f70a00>, '=', <__main__.EvalConstant object at 0x16556a430>]
```

a) Now let's copy and replace a value:
```python
parsed_copy = parsed_.copy()
parsed_copy[0].value[0] = 1
print("copy", parsed_copy[0].value.dump())
print("orig", parsed_[0].value.dump())
```
Output:
```python
copy [1, '=', <__main__.EvalConstant object at 0x16556a430>]
orig [1, '=', <__main__.EvalConstant object at 0x16556a430>]
```

I expected the original to not change, since I made a copy. 

b) So, I tried copying the top-level and inner ParseResults before making changes:

```python
parsed_ = comp_expr.parse_string("1 * 2 = 2")
parsed_copy = parsed_.copy()
parsed_copy[0].value = parsed_[0].value.copy()
parsed_copy[0].value[0] = 1
print("copy", parsed_copy[0].value.dump())
print("orig", parsed_[0].value.dump())
```
Output:
```python
copy [1, '=', <__main__.EvalConstant object at 0x165f51c10>]
orig [1, '=', <__main__.EvalConstant object at 0x165f51c10>]
```

c) Then I tried just working on the inner-most ParseResults:
```python
parsed_ = comp_expr.parse_string("1 * 2 = 2")
parsed_copy= parsed_[0].value.copy()
parsed_copy[0] = 1
print("copy", parsed_copy.dump())
print("orig", parsed_[0].value.dump())
```
Output:
```python
copy [1, '=', <__main__.EvalConstant object at 0x165fa6310>]
orig [<__main__.EvalMultOp object at 0x165f672b0>, '=', <__main__.EvalConstant object at 0x165fa6310>]
```

Success! But now I have lost the outermost object (`EvalComparisonOp`), meaning I've lost the corresponding parse actions. 

d) So, let's try changing it and feeding it back into a copy:

```python
parsed_ = comp_expr.parse_string("1 * 2 = 2")
parsed_outer_copy = parsed_.copy()
parsed_outer_copy[0].value = parsed_[0].value.copy()
parsed_inner_copy = parsed_outer_copy[0].value.copy()
parsed_inner_copy[0] = 1

# Expected: only the inner copy has changed 
print("inner_copy", parsed_inner_copy.dump())
print("outer_copy", parsed_outer_copy[0].value.dump())
print("orig", parsed_[0].value.dump())
```
Output:
```python
inner_copy [1, '=', <__main__.EvalConstant object at 0x165f67be0>]
outer_copy [<__main__.EvalMultOp object at 0x16581a040>, '=', <__main__.EvalConstant object at 0x165f67be0>]
orig [<__main__.EvalMultOp object at 0x16581a040>, '=', <__main__.EvalConstant object at 0x165f67be0>]
```
e) Now I try injecting the "inner" copy back into the "outer" copy:
```python
parsed_outer_copy[0].value = parsed_inner_copy

# Expected: only the outer copy is updated
print("inner_copy", parsed_inner_copy.dump())
print("outer_copy", parsed_outer_copy[0].value.dump())
print("orig", parsed_[0].value.dump())
```
Output
```python
inner_copy [1, '=', <__main__.EvalConstant object at 0x165f67be0>]
outer_copy [1, '=', <__main__.EvalConstant object at 0x165f67be0>]
orig [1, '=', <__main__.EvalConstant object at 0x165f67be0>]
```

Pyparsing version: 3.0.9
Python version: 3.9

**Comment by ptmcg** [comment]
Thanks for submitting this detailed scenario to replicate your issue. Yes, ParseResults.copy() is just a shallow copy. If necessary, I'll add a deepcopy() method.

**Comment by brynpickering** [comment]
Thanks for the update; makes sense. Is case `c)` above expected behaviour given that the copies are only shallow?

**Comment by ptmcg** [comment]
I'm adding a deepcopy() method in the next release, which will be an alpha - please try it out when you can to see if it does what you need.

**Comment by ptmcg** [comment]
Released in 3.1.0a1
